smart-py/marketplace.py

```diff
@@ -1,64 +1,203 @@
-        self.fa2_transfer(self.data.objkt, sp.sender, sp.self_address, params.objkt_id, params.objkt_amount)
-        self.data.swaps[self.data.counter] = sp.record(issuer=sp.sender, objkt_amount=params.objkt_amount, objkt_id=params.objkt_id, xtz_per_objkt=params.xtz_per_objkt, royalties=params.royalties, creator=params.creator)
+        # previous for easy comparison
+        # sp.verify((params.objkt_amount > 0) & ((params.royalties >= 0) & (params.royalties <= 250)))
+
+        sp.verify(
+            (params.objkt_amount > 0) &
+            (params.objkt_id < self.data.counter) &
+            (params.objkt_id >= 152) &
+            (
+                (params.royalties >= 0) & (params.royalties <= 250)
+            )
+        )
+
+        # TODO
+        # i want to validate the royalties here to make sure they match
+        # what is set on the objkt when it is minted
+
+        # previous for easy comparison
+        # self.fa2_transfer(self.data.objkt, sp.sender, sp.self_address, params.objkt_id, params.objkt_amount)
+
+        self.fa2_transfer(
+            self.data.objkt,
+            sp.sender,
+            sp.self_address,
+            params.objkt_id,
+            params.objkt_amount
+        )
+
+        # previous for easy comparison
+        # self.data.swaps[self.data.counter] = sp.record(issuer=sp.sender, objkt_amount=params.objkt_amount, objkt_id=params.objkt_id, xtz_per_objkt=params.xtz_per_objkt, royalties=params.royalties, creator=params.creator)
+
+        self.data.swaps[self.data.counter] = sp.record(
+            issuer=sp.sender,
+            objkt_amount=params.objkt_amount,
+            objkt_id=params.objkt_id,
+            xtz_per_objkt=params.xtz_per_objkt,
+            royalties=params.royalties,
+            creator=params.creator
+        )
+
         self.data.counter += 1
-    
+
     @sp.entry_point
     def collect(self, params):
         sp.verify(
             # verifies if tez amount is equal to price per objkt
-            (sp.amount == sp.utils.nat_to_mutez(sp.fst(sp.ediv(self.data.swaps[params.swap_id].xtz_per_objkt, sp.mutez(1)).open_some()))) & (self.data.swaps[params.swap_id].objkt_amount != 0))
+            (
+                sp.amount == sp.utils.nat_to_mutez(
+                    sp.fst(
+                        sp.ediv(
+                            self.data.swaps[params.swap_id].xtz_per_objkt,
+                            sp.mutez(1)
+                        ).open_some()
+                    )
+                )
+            # and verify there are objkts available for swap
+            ) & (
+                self.data.swaps[params.swap_id].objkt_amount > 0
+            # and not collecting own swap
+            ) & (
+                sp.sender != self.data.swaps[params.swap_id].issuer
+            )
+        )
 
         sp.if (self.data.swaps[params.swap_id].xtz_per_objkt != sp.tez(0)):
+            self.amount = sp.fst(
```

Comment:
The verify logic above also has "sp.amount == sp.utils.nat_to_mutez(". Not sure if intentional for the assignment here to be different.

Reply:
I also don't know why is necessary the use of sp.ediv here. xtz_per_objkt is already of type mutez (see constructor).
The code here should be:
self.amount = self.data.swaps[params.swap_id].xtz_per_objkt
unless I'm missing something

Reply:
And the same in the verify above

Reply:
are sp.amount and self.amount the same thing?